### ISSUE TITLE
IE scrolls top of iframe into view when modal within the iframe is shown

### ISSUE THREAD

**Issue by danschumann** [twbs/bootstrap#16050]
Hello,

I have an iframe that is like 3000+ pixels in height.  The top page scrolls, as the iframe will be whatever size it needs to be, so that it won't have a scrollbar.

There is a problem in IE, caused by this line: 

```
      transition ?
        that.$element.find('.modal-dialog') // wait for modal to slide in
          .one('bsTransitionEnd', function () {
            that.$element.trigger('focus').trigger(e) // this line causes the problem!!!
            that.$element.trigger(e) // this fixes the problem!!!
          })
          .emulateTransitionEnd(Modal.TRANSITION_DURATION) :
        that.$element.trigger('focus').trigger(e)
```

At approximately line 998 in 3.3.2

As you can see, removing trigger('focus') fixes it.

The problem in detail is:  the top frame's scrollTop becomes the iframes offset().top.

This can be very unwanted.  The user may be 2000 pixels into the iframe content, looking at something at the bottom, and when they open a modal, they are way at the top of the page again ( top of the iframe at least ).

Here is a hack I am using to fix:
```
    $modal.on('show.bs.modal', function(){
      // IE Fix:  bootstrap does a focus, which screws up the iframe by scrolling the parent window
      setTimeout(function(){
        $modal.find('.modal-dialog').off('bsTransitionEnd');
      }, 350);
    });
```

**Comment by cvrebert** [comment]
* What version of IE?
* Please make an example [Plunk](http://plnkr.co) that demonstrates the problem.

**Comment by danschumann** [comment]
I'm using ie11

http://plnkr.co/edit/0sLCVpuZjKOVdkCLNk7v?p=preview

to reproduce: Open any of the modals in the iframe.  

The problem: In chrome, the parent's scrollTop doesn't change.  In IE, it does ( and it was time consuming to debug ).

NOTE:  In both browsers, the modal appears at the top.  Fine.  I am listening to `parent.$(window).on('scroll resize', ...)` and setting the `$modal.css({top: ...})` accordingly.  It actually looks pretty sweet with a transition.  I know this issue is because it sees the whole iframe as a gigantic window, that's fine.

**Comment by twbs-lmvtfy** [comment]
Hi @danschumann!

You appear to have posted a live example (http://run.plnkr.co/plunks/0sLCVpuZjKOVdkCLNk7v/), which is always a good first step. However, according to [the HTML5 validator](http://html5.validator.nu), **your example has some validation errors**, which might potentially be causing your issue:
* line 1, column 33: Start tag seen without seeing a doctype first. Expected ```` <!DOCTYPE html> ````.

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

**Comment by danschumann** [comment]
http://plnkr.co/edit/0sLCVpuZjKOVdkCLNk7v?p=preview

**Comment by twbs-lmvtfy** [comment]
Hi @danschumann!

You appear to have posted a live example (http://run.plnkr.co/plunks/0sLCVpuZjKOVdkCLNk7v/), which is always a good first step. However, according to [Bootlint](https://github.com/twbs/bootlint), **your example has some Bootstrap usage errors**, which might potentially be causing your issue:
* [W001](https://github.com/twbs/bootlint/wiki/W001): `<head>` is missing UTF-8 charset `<meta>` tag
* [W002](https://github.com/twbs/bootlint/wiki/W002): `<head>` is missing X-UA-Compatible `<meta>` tag that disables old IE compatibility modes
* [W003](https://github.com/twbs/bootlint/wiki/W003): `<head>` is missing viewport `<meta>` tag that enables responsiveness

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

**Comment by danschumann** [comment]
http://plnkr.co/edit/0sLCVpuZjKOVdkCLNk7v?p=preview

**Comment by cvrebert** [comment]
Confirmed in IE11 on Windows 8.1. Also confirmed that IE11's behavior does not match Chrome, Firefox, or OS X Safari.

**Comment by cvrebert** [comment]
So, it turns out that the whole "focusing an element causes it to be scrolled into view" thing [doesn't seem to have been standardized yet](https://www.w3.org/Bugs/Public/show_bug.cgi?id=27913), which helps explain why IE11 acts differently here.
Yay Web "Standards"™!

I would actually naively argue that IE's behavior seems to make the most sense here. If something gets focused, the user should (normally) be able to see it.
Perhaps we should toss in a call to something equivalent to the non-standard `Element. scrollIntoViewIfNeeded()` method after we trigger focus on the modal?

**Comment by danschumann** [comment]
Whatever is done, I think it should be optional or opt-out-able, since if i wasn't using the transitions here, I don't think I'd have any recourse to call `.off('bsTransitionEnd')`

I don't actually want to focus, since I have to handle that myself by resizing the iframe if the modal is too big, and more commonly, offsetting the modal to match the screen.

My `modal.css({top: ...});` functions similar to the columns on facebook.  It keeps it at the top if you're above it, you can scroll through it, and when you reach the end, it keeps the bottom if you're below it.  It looks rather nice, especially considering it is an iframe workaround and not a feature.

**Comment by cvrebert** [comment]
Does the way the modal re-focuses the button that triggered it after the modal is closed also affect your use case?

**Comment by cvrebert** [comment]
@danschumann Does the way the modal re-focuses the button that triggered it after the modal is closed also affect your use case?

**Comment by fat** [comment]
we actually moved this code into the `enforceFocus` method a while ago for odd cases like this, where people need to manage their own focus state. 

To turn off bootstrap's focusing/or to implement your own, just do this:

```js
$.fn.modal.Constructor.prototype.enforceFocus = function () {}
```

**Comment by fat** [comment]
the focusing was added in a accessibility workshop - curious @patrickhlauke's feeling about it.

If we should make it optional or if their is a better api we could be implementing here

**Comment by fat** [comment]
Meanwhile updating this to a docs thing, because i think this issue has been solved before

**Comment by cvrebert** [comment]
@fat The problematic code here isn't part of `enforceFocus`, it's just part of `show`: https://github.com/twbs/bootstrap/blob/d87b7a037bbfa74fb75b6ab845803437991ec598/js/modal.js#L102

**Comment by dcousineau** [comment]
I just wanted to roll into this issue and confirm this happens with us in **Chrome on OSX**. Commenting out the `.trigger('focus')` at the very end solves the issue.

In our case we are (forced) into nested modals. We control the very inner modal but have no influence over the outer modals. The bug cropped up when the outermodal (with overflow hidden) scrolls valuable content out of view when a modal opens in our iframe.

We have a [gist reproduction here](https://gist.github.com/dcousineau/581c572bf5432e2f60e4) and a [live example](http://rawgit.com/dcousineau/581c572bf5432e2f60e4/raw/test_outer.html) of said gist in raw git.

You'll notice in the live example clicking the large or small modal buttons on the getbootstrap site loaded into the inner iframe will scroll the red content area out of view.

Our patch that fixed the problem was the following in [modal.js](https://github.com/twbs/bootstrap/blob/master/js/modal.js#L99-L105)

```javascript
        transition ?
            that.$dialog // wait for modal to slide in
                .one('bsTransitionEnd', function () {
                    that.$element/*.trigger('focus')*/.trigger(e)
                })
                .emulateTransitionEnd(Modal.TRANSITION_DURATION) :
            that.$element/*.trigger('focus')*/.trigger(e)
```

**Comment by twbs-lmvtfy** [comment]
Hi @dcousineau!

You appear to have posted a live example (https://gist.github.com/dcousineau/581c572bf5432e2f60e4), which is always a good first step. However, according to [the HTML5 validator](http://html5.validator.nu), **your example has some validation errors**, which might potentially be causing your issue:
* line 25, column 6 thru line 25, column 162: Self-closing syntax (```` /> ````) used on a non-void HTML element. Ignoring the slash and treating as a start tag.
* line 28, column 7: End of file seen when expecting text or an end tag.
* line 25, column 6 thru line 25, column 162: Unclosed element ```` iframe ````.
* line 24, column 2 thru line 24, column 71: Unclosed element ```` div ````.
* line 28, column 7: End of file seen and there were open elements.

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

**Comment by fat** [comment]
oh my bad – you're right, different accessibility focus thing - we could probably move that into it's own function as well

**Comment by cvrebert** [comment]
This is fixed in v4 via 7ef0e52fd042da2fdf107a3347abab3486a67790, which adds a new option to control the focusing logic in question.
@fat Gonna backport that to v3, or do we call this WontFix since the patch is feature-y?

**Comment by phammer** [comment]
+1 for a patch of v3

**Comment by cvrebert** [comment]
Deeming this to be a feature request, and closing as WontFix since Bootstrap v3 is essentially feature-frozen at this point. Sorry dudes.

**Comment by phammer** [comment]
@cvrebert : I agree that the proposed change is a feature because of introducing a new option. However the issue per se is a bug in bs3 for me because the behavior isn't consistent.
